hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java

```diff
@@ -257,33 +266,39 @@ private void removeExpiredTransactions(
-
-        // Loop in order and expunge every entry where the timestamp is before the current time. Also remove from the
-        // in memory data structures.
-        final var itr = queue.iterator();
-        while (itr.hasNext()) {
-            final var entry = itr.next();
-            final var rec = entry.transactionRecordOrThrow();
-            final var txId = rec.transactionIDOrThrow();
-            // If the timestamp is before the current time, then it has expired
-            if (isBefore(txId.transactionValidStartOrThrow(), earliestValidState)) {
-                // Remove from the histories
-                itr.remove();
-                // Remove from the payer to transaction index
-                final var payerAccountId = txId.accountIDOrThrow(); // NOTE: Not accurate if the payer was the node
-                final var transactionIDs =
-                        payerToTransactionIndex.computeIfAbsent(payerAccountId, ignored -> new HashSet<>());
-                transactionIDs.remove(txId);
-                if (transactionIDs.isEmpty()) {
-                    payerToTransactionIndex.remove(payerAccountId);
+        final var earliestValidStart = minus(consensusTimestamp, config.transactionMaxValidDuration());
+        // Loop in order and expunge every entry where the start time is before the earliest valid start.
+        // Also remove from the in-memory data structures.
+        do {
+            final var entry = queue.peek();
+            if (entry != null) {
+                final var rec = entry.transactionRecordOrThrow();
+                final var txId = rec.transactionIDOrThrow();
+                // If the valid start time is before the earliest valid start, then it has expired
+                if (isBefore(txId.transactionValidStartOrThrow(), earliestValidStart)) {
+                    // Remove from the histories.  Note that all transactions are added to this map
+                    // keyed to the "user transaction" ID, so removing the entry here removes both
+                    // "parent" and "child" transaction records associated with that ID.
+                    histories.remove(txId);
+                    // remove from queue as well.  The queue only permits removing the current "HEAD",
+                    // but that should always be correct here.
+                    queue.removeIf(TruePredicate.INSTANCE);
+                    // Remove from the payer to transaction index
+                    final var payerAccountId = txId.accountIDOrThrow(); // NOTE: Not accurate if the payer was the node
+                    final var transactionIDs =
+                            payerToTransactionIndex.computeIfAbsent(payerAccountId, ignored -> new HashSet<>());
+                    transactionIDs.remove(txId);
+                    if (transactionIDs.isEmpty()) {
+                        payerToTransactionIndex.remove(payerAccountId);
+                    }
```

Comment:
The loop updates `histories`, the queue itself, and `payerToTransactionIndex`, but does nothing with the deduplication cache. Should the transaction ID also be removed from there?

Reply:
If it was used, probably.  At the moment the _use_ of the deduplication cache is not implemented (and it's not certain it needs to be; it might be removed in a future PR).